### PR TITLE
#820 attribute value code list concept scheme v2

### PR DESCRIPTION
**Pull request by SarahJohnsonONS**
Ticket [#820](https://app.zenhub.com/workspaces/features-squad-61b72275ac896c0010a1b00b/issues/gh/gss-cogs/csvcubed/820)

In order to allow for old schema versions to continue to work, the approach taken was to replace the `new_attribute_values` property with a `code_list` property and amend the existing code so that the correct RDF triples are generated in the data structure definition from the Attribute codelist and its concepts.

In order for the new functionality to work, the Attribute column in the CSV file must be dense (i.e. a value in every cell). If a user attempts to build a cube with attribute value codelists, and the relevant Attribute column has missing values, this will raise an error.

A feature flag has been created to allow for the new functionality to be toggled on and off - see `src/csvcubed/feature_flags.py`. To see how this flag has been implemented, replace "false" with "true" (string value) in `os.environ.get("OUTPUT_ATTR_VAL_CODE_LISTS", "false")`

The main changes made are:

#### models/cube/qb/components/attribute.py

- `QbAttribute.get_new_attribute_values()` abstract method removed - only New Resource Attributes now have values associated with them, so this no longer needs to be an abstract method
- `new_attribute_values` removed from `ExistingQbAttribute` fields
- `new_attribute_values` replaced with `code_list` in `NewQbAttribute` fields with an `Optional[QbCodeList]` type
- `NewQbAttribute.from_data()` now generates a codelist from the column data if ATTRIBUTE_VALUE_CODELISTS is True, otherwise a list of `NewQbConcept`s  must be provided, generated from the `values` object in the config.json.
- `NewQbAttribute._validate_concepts_non_conflicting` removed as this check is performed in NewQbCodelist

#### models/rdf/newattributevalueresource.py
`CodedAttributeProperty` class added to allow for Attributes with codelists. This isn't part of the formal QB specification so needs to be defined outside of the csvcubedmodels repo.

#### readers/cubeconfig/v1/columnschema.py
`ExistingAttributeResource.map_to_existing_qb_attribute` changed to `map_to_qb_attribute`. If a `config.json` file matches the fields for an `ExistingAtributeResource`, and `values` is set to `True` or a list of `AttributeValues`, this is now mapped to a `NewQbAttribute` with a codelist generated from the values in the CSV column or the list of defined values respectively.

#### writers/helpers/qbwriter/dsdtordfmodelshelper.py
`DsdToRdfModelsHelper.generate_data_structure_definitions` now generates different results depending on whether `ATTRIBUTE_VALUE_CODELISTS` is set to `True` or `False`:
If False:
- attribute value resources are added to the rdfs:seeAlso object in the main csv-metadata.json
If True:
- URIRefs are generated to link the codelist files to the Attribute component in the main csv-metadata.json
- the Attributes are included in the file dependencies

#### writers/helpers/qbwriter/urihelper.py
`UriHelper._get_default_property_value_uris_for_attribute` generates different valueUrls depending on whether `ATTRIBUTE_VALUE_CODELISTS` is set to `True` or `False`:
If True:
- valueUrl is 'attribute.csv#{+attribute}'
If False:
- valueUrl is 'cube-title.csv#attribute/attribute/{+attribute}'

#### writers/qbwriter.py
`QbWriter.write` updated to output Attribute codelist CSVWs if `ATTRIBUTE_VALUE_CODELISTS` is set to `True` (this creates the attribute.csv, attribute.csv-metadata.json and attribute.table.json files)
Attributes are also included in the foreign keys if ATTRIBUTE_VALUE_CODELISTS is True

### CODE REVIEW

File: src/csvcubed/writers/qbwriter.py [hunk 1]
```diff
@@ -100,15 +105,17 @@ def write(self, output_folder: Path):
...
-        self._output_new_code_list_csvws(output_folder)
-
         csvw_metadata = {
             "@context": "http://www.w3.org/ns/csvw",
             "@id": self._uris.get_dataset_uri(),
             "tables": tables,
             "rdfs:seeAlso": self._dsd.generate_data_structure_definitions(),
         }
 
+        self._output_new_dimension_code_list_csvws(output_folder)
```
Comment (canwaf) [inline src/csvcubed/writers/qbwriter.py:115]: Can we use the same function for dimension and attribute code lists?

File: tests/behaviour/steps/qbwriter.py [hunk 2]
```diff
@@ -101,8 +106,10 @@ def step_impl(context, cube_name: str):
         QbColumn(
             "Marker",
             NewQbAttribute.from_data(
+                "Marker",
                 "Marker",
                 data["Marker"],
+                [NewQbConcept("Suppressed")],
```
Comment (canwaf) [inline tests/behaviour/steps/qbwriter.py:112]: Make this not written out manually. e.g. data["Marker"].dropna() and as a list comprenension.

`[NewQbAttribute(value) for value in data["Marker"].dropna()]`

File: tests/test-cases/readers/cube-config/v1.0/attribute_value_codelists.csv
Comment (canwaf) [inline tests/test-cases/readers/cube-config/v1.0/attribute_value_codelists.csv:1]: Delete?

File: tests/test-cases/readers/cube-config/v1.0/attribute_value_codelists.json
Comment (canwaf) [inline tests/test-cases/readers/cube-config/v1.0/attribute_value_codelists.json:1]: Delete?

File: tests/unit/cube/qb/test_cubeqb_errorvalidation.py [hunk 3]
```diff
@@ -918,7 +1057,9 @@ def test_pivoted_validation_multiple_measure_columns():
     assert isinstance(error, PivotedShapeMeasureColumnsExistError)
 
 
-def test_pivoted_validation_no_measures_defined_error():
+def test_pivoted_validation_no_measures_defined_error_codelist_true(
```
Comment (canwaf) [inline tests/unit/cube/qb/test_cubeqb_errorvalidation.py:1060]: You went overboard on this. Our testing is that no measure is failing. Delete the duplicates tests.

File: tests/unit/cube/qb/test_cubeqb_errorvalidation.py
Comment (canwaf) [inline tests/unit/cube/qb/test_cubeqb_errorvalidation.py:1]: There are a number of tests with no interactions between attribute value code lists and the error being expected. We *should* assume that measure dimensions have no impact on attribute value code lists being generated or not, if we get another error in a test case, great, the test fails and we need to debug it test purpose accomplished.

File: tests/behaviour/qbwriter.feature [hunk 4]
```diff
@@ -404,12 +449,55 @@ Feature: Test outputting CSV-Ws with Qb flavouring.
...
+    Given the ATTRIBUTE_VALUE_CODELISTS feature flag is set to False
+    And a single-measure QbCube named "Bad Qube" with missing observation values and missing `sdmxa:obsStatus` replacements
+    Then the CSVqb should fail validation with "Missing value(s) found for 'Value' in row(s) 0"
+    When the cube is serialised to CSV-W
+    # Unfortunately, CSV-W validation will *not* catch this error since the obs column cannot be marked as `required`
+    # since an `sdmxa:obsStatus` Attribute column has been defined.
+    Then csvlint validation of "bad-qube.csv-metadata.json" should succeed
+
+# TODO 820 Fix this test
+  # Scenario: A cube with an optional attribute which has missing data values should validate successfully given ATTRIBUTE_VALUE_CODELISTS is True
```
Comment (canwaf) [inline tests/behaviour/qbwriter.feature:476]: Delete this, attribute code lists require a dense attribute code list.

File: tests/behaviour/qbwriter.feature [hunk 5]
```diff
@@ -404,12 +449,55 @@ Feature: Test outputting CSV-Ws with Qb flavouring.
...
+  #   Given the ATTRIBUTE_VALUE_CODELISTS feature flag is set to True
+  #   And a single-measure QbCube named "Some Qube" with optional attribute values missing
+  #   Then the CSVqb should pass all validations
+  #   When the cube is serialised to CSV-W
+  #   Then csvlint validation of all CSV-Ws should succeed
+
+# TODO 820 Fix this test
```
Comment (canwaf) [inline tests/behaviour/qbwriter.feature:483]: Follow up this ticket with a new one, which makes sure we support a sparse attribute column (i.e. obsStatus) when we have an attribute column with a code list (and thus is dense).

(This test can obvs go.)

File: tests/behaviour/qbwriter.feature [hunk 6]
```diff
@@ -404,12 +449,55 @@ Feature: Test outputting CSV-Ws with Qb flavouring.
...
+  # Scenario: Observation Values are Optional where an 'sdmxa:ObsStatus' Attribute is Present given ATTRIBUTE_VALUE_CODELISTS is True
+  #   Given the ATTRIBUTE_VALUE_CODELISTS feature flag is set to True
+  #   And a single-measure QbCube named "Good Qube" with missing observation values and `sdmxa:obsStatus` replacements
+  #   Then the CSVqb should pass all validations
+  #   When the cube is serialised to CSV-W
+  #   Then csvwcheck validation of "good-qube.csv-metadata.json" should succeed
+
+# TODO 820 Fix this test
```
Comment (canwaf) [inline tests/behaviour/qbwriter.feature:491]: Delete. Delete Delete. Delete. Block.

File: tests/unit/utils/qb/validation/test_cube.py [hunk 7]
```diff
@@ -43,9 +44,9 @@ def test_new_qb_attribute_literal_string_with_template():
     )
 
 
-def test_existing_qb_attribute_resource_no_values():
+def test_new_qb_attribute_resource_no_values():
```
Comment (canwaf) [inline tests/unit/utils/qb/validation/test_cube.py:47]: Return this test, we should still support existing qb attribute resource with no values (e.g. a csv-w with observations not missing but configured to support observations missing with an obsStatus value)

File: tests/unit/writers/qbwriter/test_qbwriter.py [hunk 8]
```diff
@@ -119,7 +213,7 @@ def test_output_new_code_list_csws_urls_with_uri_style_without_file_extensions()
     qb_writer = QbWriter(cube)
     with TemporaryDirectory() as temp_dir:
         temp_dir = Path(temp_dir)
-        qb_writer._output_new_code_list_csvws(temp_dir)
+        qb_writer._output_new_dimension_code_list_csvws(temp_dir)
```
Comment (canwaf) [inline tests/unit/writers/qbwriter/test_qbwriter.py:216]: Double check if we can consolidate the dimension_code_list and attribute_code_list functions.

File: tests/unit/writers/qbwriter/test_urihelper.py [hunk 9]
```diff
@@ -235,17 +236,14 @@ def test_default_property_value_uris_existing_attribute_new_values():
     """
     column = QbColumn(
         "Some Column",
-        ExistingQbAttribute(
-            "http://base-uri/attributes/existing-attribute",
-            new_attribute_values=[NewQbAttributeValue("Some Attribute Value")],
-        ),
+        ExistingQbAttribute("http://base-uri/attributes/existing-attribute"),
```
Comment (canwaf) [inline tests/unit/writers/qbwriter/test_urihelper.py:239]: Restore this test.

**Review by canwaf — CHANGES_REQUESTED** [review]
Excellent progress on attribute code lists.

File: tests/behaviour/qbwriter.feature (hunk 5 above)
Comment (SarahJohnsonONS) [reply tests/behaviour/qbwriter.feature:483]: Created new ticket (#873) and added to epic

File: tests/unit/utils/qb/validation/test_cube.py (hunk 7 above)
Comment (SarahJohnsonONS) [reply tests/unit/utils/qb/validation/test_cube.py:47]: I've left this one as-is, as it's properly testing that the NoUriTemplateOrAttrValuesError is being raised. I've added a new test (test_existing_qb_attribute_resource_no_cell_uri_template), and a new elif to the _validate_attributes function in utils/qb/validation/cube.py to check that an ExistingQbAttribute with no cell_uri_template raises CsvColumnUriTemplateMissingError

**Review by NickPapONS — APPROVED** [review]
I have a couple of questions about some parts but I think the previous review got pretty much everything that needed real changes requested. I will give approval since the stuff I found is incredibly minor, so feel free to proceed after making the couple of changes/answering my questions. Overall you did great work on this and also included some very useful comments everywhere, I have a decent understanding of how all these many parts work now.

File: tests/behaviour/steps/qbwriter.py [hunk 10]
```diff
@@ -68,19 +69,23 @@ def get_standard_catalog_metadata_for_name(
...
 
 @Given('a single-measure QbCube named "{cube_name}" with missing observation values')
 def step_impl(context, cube_name: str):
+    feature_flags.ATTRIBUTE_VALUE_CODELISTS = context.flag
     cube = _get_single_measure_cube_with_name_and_id(cube_name, None)
     cube.data["Value"] = [1, None, 3]
     context.cube = cube
 
 
+# qbwriter.feature 484
```
Comment (NickPapONS) [inline tests/behaviour/steps/qbwriter.py:83]: Leftover comment here

File: src/csvcubed/models/cube/qb/components/attribute.py [hunk 11]
```diff
@@ -203,16 +175,24 @@ def validate_data(
...
+        if (
+            isinstance(self.code_list, NewQbCodeList)
+            and len(self.code_list.concepts) > 0
+        ):
+            expected_values = {concept.code for concept in self.code_list.concepts}
+            actual_values = {
+                uri_safe(str(v)) for v in set(data.unique()) if not pd.isna(v)
+            }
+            undefined_values = actual_values - expected_values
+            if len(undefined_values) > 0:
```
[the remark this answers is not in the document]
Comment (SarahJohnsonONS) [reply src/csvcubed/models/cube/qb/components/attribute.py:187]: Yes but that's OK from a codelist perspective, i.e. an attribute value is defined in the config (expected value) but isn't used in the attribute column (actual value), that's acceptable.

File: src/csvcubed/utils/qb/validation/cube.py [hunk 12]
```diff
@@ -62,20 +67,25 @@ def _validate_attributes(cube: Cube) -> List[ValidationError]:
...
                     )
-            else:
-                # Not a QbAttributeLiteral
-                if (
-                    c.csv_column_uri_template is None
-                    and len(c.structural_definition.new_attribute_values) == 0  # type: ignore
-                ):
+            # Not a QbAttributeLiteral
+            elif isinstance(c.structural_definition, NewQbAttribute):
+                if c.csv_column_uri_template is None and c.structural_definition.code_list is None:  # type: ignore
```
[the remark this answers is not in the document]
Comment (SarahJohnsonONS) [reply src/csvcubed/utils/qb/validation/cube.py:75]: It was to fix a Pyright error

File: tests/behaviour/steps/qbwriter.py (hunk 10 above)
Comment (SarahJohnsonONS) [reply tests/behaviour/steps/qbwriter.py:83]: Will delete

File: src/csvcubed/readers/cubeconfig/v1/columnschema.py [hunk 13]
```diff
@@ -356,10 +381,10 @@ def map_to_new_qb_attribute(self, column_title: str) -> NewQbAttributeLiteral:
...
     label: Optional[str] = None
     description: Optional[str] = None
-    from_existing: Optional[str] = None
     definition_uri: Optional[str] = None
-    required: bool = False
+    from_existing: Optional[str] = None
     values: Union[bool, List[AttributeValue]] = True
+    required: bool = False
     cell_uri_template: Optional[str] = None
     describes_observations: Optional[str] = None
```
[the remark this answers is not in the document]
Comment (SarahJohnsonONS) [reply src/csvcubed/readers/cubeconfig/v1/columnschema.py:390]: I tried but failed XD

**Review by canwaf — APPROVED** [review]

**Pull request merged**